iwyu.cc

```diff
@@ -4313,6 +4338,25 @@ class IwyuAstConsumer
     return false;
   }
 
+  bool SourceOrTargetTypeIsProvided(const ASTNode* construct_expr_node) const {
+    const auto* expr = construct_expr_node->GetAs<CXXConstructExpr>();
```

Comment:
Can you add a `CHECK_` here that this return non-null, and so is actually a `CXXConstructExpr`?

Reply:
Btw, I remember you had the opposite opinion: https://github.com/include-what-you-use/include-what-you-use/pull/1059#discussion_r884479721

Reply:
Yeah, I had a feeling, thanks for looking it up. It's somewhat arbitrary, but here's my reasoning:

* `const EnumDecl* decl = type->getDecl();` should normally never fail; we have strong types and a reasonable assumption about how things work
* in this method, we have zero support from the type system, the only hint is the name of the parameter. Adding a `CHECK_` here is executable documentation that `construct_expr_node` MUST be a `CXXConstructExpr`. It might be nicer with `CHECK_(node->IsA<CXXConstructExpr>());` up-front, but then we have to cast twice.